### DISCUSSION TITLE
Single threaded client application

### DISCUSSION THREAD

**Discussion by nickjmeyer** [aws/aws-iot-device-sdk-cpp-v2#718] [Q&A] [answered]
According to the [documentation](https://docs.aws.amazon.com/greengrass/v2/developerguide/interprocess-communication.html#ipc-client-versions) about the changes between V1 and V2, it sounds like all subscriptions have their own thread.

Is it possible to keep the entire event loop on a single thread?

**Comment by bretambrose** [comment, marked as answer]
v1 and v2 are a bit overloaded terms, unfortunately.  Within the v2 device SDKs, GG IPC has two different clients:

v1 IPC clients - callbacks all occur on the event loop thread that the network connection is bound to.  It was a common occurrence for developers to accidentally deadlock their components by waiting on a completion event within a callback.  This led the Greengrass team to develop "v2"/higher-level IPC clients that execute callbacks on a thread pool, avoiding the deadlock scenario.

From a language-SDK perspective:
* C++ - only has the v1 client.  Overall, can be difficult to use and we'd like to improve the experience but haven't had time yet.  To answer your question directly, C++ IPC callbacks only occur on a single thread -- the thread the connection is bound to.
* Java - both v1 and v2 IPC clients
* Python - both v1 and v2 IPC clients
* JS - v2-esque client (doesn't run on a thread pool, just integrates naturally into node's event loop)

**Comment by nickjmeyer** [comment]
Gotcha.  Thanks! The explanation is much appreciated.
